jbpm-persistence/jbpm-persistence-jpa/src/main/java/org/jbpm/persistence/processinstance/JPAProcessInstanceManager.java

```diff
@@ -126,6 +130,9 @@ public ProcessInstance getProcessInstance(long id, boolean readOnly) {
                 }  
                 TransactionManagerHelper.addToUpdatableSet(txm, processInstanceInfo);
                 processInstanceInfo.updateLastReadDate();
+                
+
+                EventManagerProvider.getInstance().get().update(new ProcessInstanceView(processInstance));
   
```

Comment:
Just a question on how update of ProcessInstanceInfo works. I see that it is added to updatable set and then somehow I guess a transform method is called and it is persisted, right? Does this happen automatically with any ProcessInstance change after every commit? I was quite surprised that actually a getProcessInstance method updates a ProcessInstance in DB. And PersistenceEventManager will be notified about update only when getProcessInstance is called which may be much later I guess. Imagine that the ProcessInstance is updated by the engine, e.g. Human Task is encountered. Now we wait for a Human Task to be completed, let's say it is completed after 5 minutes - I guess only then is getProcessInstance called again so for the PersistenceEventManager  the process is updated only after that, right? But I am only guessing how this works, of course :)

Reply:
update to process instance is done only once upon transaction completion (using TransactionSychronization.beforeCompletion). This is when every process instance that was loaded (not in read only mode) will be persisted back to data base - and this is retrieved from updatable set. So getProcessInstance is not updating anything, it's just marking it to be persisted when tx completes.

Reply:
OK, understand now. So EventManager is just updated with a new ProcessInstanceView of a processInstance and since it is only a pointer, any change made to the processInstance after this method is called will be still recorded, right?

Reply:
yes, ProcessInstanceView has method copyFromSource and that is when it actually transforms the source (process instance) into process instance view. And this is called when collection is about to be delivered to emitter https://github.com/kiegroup/jbpm/pull/951/files#diff-ec43d650e5daf7c5415d81f2783a3f06R115